[Design system] [Design2-53] Segmented Buttons component

levadadenys opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->
### [Design system] [Design2-53] Segmented Buttons component

* Created `SegmentedButtons` component
* created stories for  `SegmentedButtons`
* created tests for  `SegmentedButtons`
* created documentation for  `SegmentedButtons`

<img width="869" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/c0c96619-8884-480d-91e9-328a798385af">

https://github.com/code-dot-org/code-dot-org/assets/22244040/bcbdf1e6-c308-4d56-bc22-2886238c8938


## Links
- jira ticket: [Design2-53](https://codedotorg.atlassian.net/browse/DESIGN2-53)



<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

sanchitmalhotra126 commented on apps/src/componentLibrary/segmentedButtons/_SegmentedButton.tsx:
```diff
@@ -0,0 +1,88 @@
+  disabled?: boolean;
+  /** Is button selected */
+  selected?: boolean;
+  /** Button unique value. Used for selected/not selected logic */
+  value: string;
```
Would `id` make more sense, since this is meant to be unique and used in selection logic?

levadadenys replied:
good question, not sure though. That's exactly the reason why I named this prop a value :D .
If we name it `id`, that might confuse someone that that stands for html `id` attribute. Meanwhile `value` is more like drpopdown option `value`, or regular button's `value` for example

sanchitmalhotra126 commented on apps/src/componentLibrary/segmentedButtons/SegmentedButtons.tsx:
```diff
@@ -0,0 +1,73 @@
+        <SegmentedButton
+          key={label}
+          selected={selectedButtonValue === value}
+          label={label}
+          onChange={onChange}
+          disabled={disabled}
+          iconLeft={iconLeft}
+          iconRight={iconRight}
+          icon={icon}
+          buttonType={type}
+          value={value}
```
nit: don't break apart the properties in the `map` function but instead spread in here in the props, like `{...button}`?

levadadenys replied:
Good question. I prefer this approach since it makes code a little more transparent, you don't need to go to the `_SegmentedButton` component to check what are the props being passed into it, one can see everything on the go.

Do we have some prediscussed rules or approaches about this? If yes I'm good to update it, if not - I would rather leave it as is for now

sanchitmalhotra126 commented on apps/src/componentLibrary/segmentedButtons/_SegmentedButton.tsx:
```diff
@@ -0,0 +1,88 @@
+  title: string;
+};
+
+export type SegmentButtonType = 'withLabel' | 'iconOnly' | 'number';
```
Should this component enforce any requirements around this type - like if `iconOnly` is selected, then `icon` must be defined, etc?

levadadenys replied:
Yeah, that would be perfect. Still looking for a way to implement it though

fisher-alice replied:
Just curious if we could create a subtype of`_SegmentedButton`, e.g., `_SegmentedButtonIconOnly` and restrict there? Not sure about terminology or possibility...

levadadenys replied:
Thanks for the hint, Alice!
Spent some time with ChatGpt and  ~~worked out~~ a solution for this.

levadadenys replied:
ok, it turned out that
1. That solution caused a lot of type errors in `SegmentedButtons.tsx` and `SegmentedButtons.story.tsx` .
I've tried to fix that, but didn't succed. I'll try again tomorrow, but not sure if that's possible. In case I won't found a quick solution for this - I suggest we revert last two commits in this PR (
[rework _segmentedButton types to be more strict and clear](https://github.com/code-dot-org/code-dot-org/pull/52357/commits/d0e313971eb995bc6c80339f99c16b32d53a13ce) and 
[a try to fix type errors](https://github.com/code-dot-org/code-dot-org/pull/52357/commits/2b1174545e58bf9755c8c99074f6189e2238308f)
) and merge this PR as it was before when this thread has started.

2. I think we should TypeScript type check as one of our pre-commit hooks, since I was able to commit files with type errors two times, one of them without knowing that I've broken anything.

fisher-alice replied:
Thanks Denys! I'm definitely new to TS so not sure how to help with fixing these type errors. So reverting the last 2 commits SGTM!

Another question I just thought of is:  are we assuming that a `SegmentedButtons` component would consist of a list of `_SegmentedButton`'s of the same `buttonType`? Or would it be okay to have a `SegmentedButtons` component which has mixed `buttonType`'s - such as an icon with a number button?

sanchitmalhotra126 replied:
Yeah doing conditional type verification can get tricky. Might be better to just have some runtime warnings or even just clearer documentation what props are required for the different button types (e.g. "If using the `withLabel` or `number` button type, the `label` prop must be provided").

fisher-alice replied:
+1 on adding clearer documentation - thanks!

levadadenys replied:
yeah, documentation should be the best option for now.
Will update, thanks guys!

levadadenys replied:
> Another question I just thought of is: are we assuming that a `SegmentedButtons` component would consist of a list of `_SegmentedButton`'s of the same `buttonType`? Or would it be okay to have a `SegmentedButtons` component which has mixed `buttonType`'s - such as an icon with a number button?

Yeah, all the buttons in `SegmentedButtons` should be of the same type, and the only way to set that type - is passing `type` prop to `SegmentedButtons` component. 

Commits that had the 'a try to make conditional types'  might have been a little bit confusing, please check again once I'll push the latest version of this pr

sanchitmalhotra126 commented on apps/src/componentLibrary/segmentedButtons/segmentedButtons.module.scss:
```diff
@@ -0,0 +1,218 @@
+  // Pressed styles
+  &:active {
+    background-color: $light_primary_700;
+    // Important here is needed to rewrite line 3320, app/assets/stylesheets/application.scss */
```
Just to check - on line [3320](https://github.com/code-dot-org/code-dot-org/blob/c7dbf69132b9de2ec7ac7e06894fd89dac2da5ee/dashboard/app/assets/stylesheets/application.scss#L3320) I see a style rule for `button>img`. Is this in reference to line [3328](https://github.com/code-dot-org/code-dot-org/blob/c7dbf69132b9de2ec7ac7e06894fd89dac2da5ee/dashboard/app/assets/stylesheets/application.scss#L3328)? To undo the `button:active border`?

Tangentially, the `!important` button styles in application.scss have caused us to create all sorts of overrides, so I'd be really curious to see if we could remove those somehow...

levadadenys replied:
Yep, it's for 3328 line, thanks for the catch!
I would like to try that, but I would rather create a ticket for that and do it as a separate ticket since that might cause a lot of eyes diffs potentially and I want this PR to only affect design system component

sanchitmalhotra126 replied:
Oh yep for sure should be a different change 🙂 I've just been seeing it pop up a lot recently.

levadadenys replied:
Created a quick win ticket for this: https://codedotorg.atlassian.net/browse/DESIGN2-88

sanchitmalhotra126 commented on apps/test/unit/componentLibrary/SegmentedButtonsTest.jsx:
```diff
@@ -0,0 +1,162 @@
+
+    expect(segmentedButton1).to.exist;
+    expect(segmentedButton2).to.exist;
+    expect(valuesMap.label || 'label').to.equal('label');
```
is the `|| 'label'` needed here? Wouldn't this make the check pass even if the callback didn't fire? (similarly for the other tests)

levadadenys replied:
Fixed, thanks

sanchitmalhotra126 commented on apps/test/unit/componentLibrary/SegmentedButtonsTest.jsx:
```diff
@@ -0,0 +1,162 @@
+    expect(valuesMap.label || 'label').to.equal('label');
+  });
+
+  it('SemgentedButtons - changes selected button on click', async () => {
```
nit: *Segmented

fisher-alice commented on apps/src/componentLibrary/segmentedButtons/_SegmentedButton.tsx:
```diff
@@ -0,0 +1,88 @@
+import moduleStyles from '@cdo/apps/componentLibrary/segmentedButtons/segmentedButtons.module.scss';
+import FontAwesomeV6Icon from '@cdo/apps/componentLibrary/fontAwesomeV6Icon';
+
+type SegmentedButtonIconProps = {
```
Is there a reason that we don't use `FontAwesomeV6IconProps` interface here?

levadadenys replied:
There're none. Thanks a lot for spotting this! Will fix

fisher-alice commented on apps/src/componentLibrary/segmentedButtons/SegmentedButtons.story.tsx:
```diff
@@ -0,0 +1,310 @@
+  components: [
+    {
+      buttons: [
+        {label: 'Label', value: 'label'},
```
Nit: 'Label one' for consistency.

levadadenys replied:
Oh, that's one is actually intended, so that one who'll be checking storybook will be able to see how one word `_SegmentedButton` will look like

fisher-alice commented on apps/src/componentLibrary/segmentedButtons/SegmentedButtons.story.tsx:
```diff
@@ -0,0 +1,310 @@
+        {
+          label: 'Label',
+          value: 'label-ir',
+          iconRight: {iconName: 'check', iconStyle: 'solid', title: 'check'},
```
You could use a different icon so folks know if just looking at the storybook that the icon is not always a checkmark when used with a text label.

levadadenys replied:
Good point, thanks. Will fix

fisher-alice reviewed (approved):
Great job! Left just a couple minor questions.
It's so cool to see `componentLibrary` growing with consistency and documentation! 🎉

levadadenys commented:
I've fixed all the comments, updated the documentation and pushed it to the RP. Once the drone will succeed - I suggest we go ahead and merge this PR.

fisher-alice commented on apps/src/componentLibrary/segmentedButtons/_SegmentedButton.tsx:
```diff
@@ -62,15 +67,15 @@ const SegmentedButton: React.FunctionComponent<SegmentedButtonProps> = ({
         />
       )}
-      {iconLeft && (
+      {buttonType === 'withLabel' && iconLeft && (
```
Nice!

fisher-alice commented on apps/src/componentLibrary/segmentedButtons/SegmentedButtons.story.tsx:
```diff
@@ -200,12 +200,12 @@ TypesOfSegmentedButtons.args = {
           label: 'Label Two',
           value: 'label-two-il',
-          iconLeft: {iconName: 'check', iconStyle: 'solid', title: 'check'},
+          iconLeft: {iconName: 'house', iconStyle: 'solid', title: 'check'},
```
Thanks for rendering different types of icons here! I think we can merge as-is, but maybe we can update `title` of these in another PR.

levadadenys replied:
Oh yeah, thanks a lot for spotting these, I forgot about them. Will update in a followup PR

fisher-alice commented:
> I've fixed all the comments, updated the documentation and pushed it to the RP. Once the drone will succeed - I suggest we go ahead and merge this PR.

LGTM - Thanks for the updates!
